tuna/connection.py

```diff
@@ -190,87 +276,46 @@ def exec_command(self, cmd, timeout=SSH_TIMEOUT, abort=None, proc_line=None):
       if not proc_line:
         # pylint: disable-next=unnecessary-lambda-assignment ; more readable
         proc_line = lambda x: self.logger.info(line.strip())
-      ret_out = StringIO()
-      ret_err = e_var
+      ret_out: Union[TextIO, StringIO] = StringIO()
+      ret_err: ChannelStderrFile = e_var
+      ret_code: int = 0
       try:
         while True:
-          line = o_var.readline()
+          line: str = o_var.readline()
           if line == '' and not self.is_alive():
             break
           if line:
             proc_line(line)
             ret_out.write(line)
         ret_out.seek(0)
         if self.local_machine:
-          ret_code = self.subp.returncode
+          if self.subp is not None:
+            ret_code = self.subp.returncode
+          else:
+            ret_code = 0
           if ret_out:
             ret_out.seek(0)
-            ret_err = StringIO()
-            err_str = ret_out.readlines()
+            ret_err = StringIO()  #type: ignore
```

Comment:
We could use another union on this variable with StringIO, then we don't need to ignore it. Eventually these should be converted to pure StringIO types.